File: src/contest/ioi/IOI_2004_Empodia.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 500100
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

int N;
int hi[SIZE], lo[SIZE], val[SIZE];

vector<vector<int>> poss (SIZE * 2);
set<pi> intervals;

struct Event {
    int x, index, type;
    Event (int x, int type, int index) {
        this->x = x;
        this->type = type;
        this->index = index;
    }

    bool operator < (const Event& e) const {
        if (x == e.x && type == e.type)
            return e.index > index;
        if (x == e.x)
            return type > e.type;
        return x > e.x;
    }
};

void sweep (vector<int> indexes) {
    priority_queue<Event> pq;
    for (int i = 0; i < indexes.size(); i++) {
        pq.push(Event(lo[val[indexes[i]]] + 1, 1, indexes[i]));
        pq.push(Event(hi[val[indexes[i]]], -1, indexes[i]));
    }

    set<int> active;
    while (!pq.empty()) {
        Event curr = pq.top();
        pq.pop();

        if (curr.type == 1)
            active.insert(curr.index);
        else if (curr.type == -1) {
            active.erase(curr.index);
            auto lower = active.upper_bound(curr.index);
            if (lower != active.begin() && (*(--lower)) > lo[val[curr.index]]) {
                pi add = mp((*lower), curr.index);
                auto prev = intervals.upper_bound(add);
                auto next = intervals.lower_bound(add);
                bool intersectPrev = true;
                bool intersectNext = true;

                if (prev != intervals.begin()) {
                    prev--;
                    if (max(add.first, (*prev).first) < min(add.second, (*prev).second)) {
                        if (add.second - add.first <= (*prev).second - (*prev).first) {
                            intervals.erase(prev);
                            intersectPrev = false;
                        }
                    } else {
                        intersectPrev = false;
                    }
                } else {
                    intersectPrev = false;
                }

                if (next != intervals.end()) {
                    if (max(add.first, (*next).first) < min(add.second, (*next).second)) {
                        if (add.second - add.first <= (*next).second - (*next).first) {
                            intervals.erase(next);
                            intersectNext = false;
                        }
                    } else {
                        intersectNext = false;
                    }
                } else {
                    intersectNext = false;
                }
                if (!intersectPrev && !intersectNext)
                    intervals.insert(add);
            }
        }
    }
}

int main () {
	// freopen("in.txt", "r", stdin);
	// freopen("out.txt", "w", stdout);

    scanf("%d", &N);

    for (int i = 0; i < N; i++) {
        scanf("%d", &val[i]);
        poss[val[i] - i + N - 1].pb(i);
    }

    // value, index
    stack<pi> s;

    for (int i = N - 1; i >= 0; i--) {
        while (!s.empty() && val[i] < s.top().first)
            s.pop();
        if (s.empty())
            hi[val[i]] = N;
        else
            hi[val[i]] = s.top().second;
        s.push(mp(val[i], i));
    }

    while (!s.empty())
        s.pop();

    for (int i = 0; i < N; i++) {
        while (!s.empty() && val[i] > s.top().first)
            s.pop();
        if (s.empty())
            lo[val[i]] = -1;
        else
            lo[val[i]] = s.top().second;
        s.push(mp(val[i], i));
    }

    for (int i = 0; i < 2 * N; i++)
        sweep(poss[i]);

    printf("%d\n", intervals.size());
    for (pi i : intervals)
        printf("%d %d\n", i.first + 1, i.second + 1);

	return 0;
}

File: src/contest/ioi/IOI_2004_Empodia.java
package contest.ioi;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.StringTokenizer;
import java.util.TreeSet;

public class IOI_2004_Empodia {

  static BufferedReader br;
  static PrintWriter out;
  static StringTokenizer st;

  static int N;
  static int[] hi, lo, val;
  static ArrayList<ArrayList<Integer>> poss;
  static TreeSet<Interval> intervals;

  public static void main (String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new PrintWriter(new OutputStreamWriter(System.out));
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    N = readInt();

    val = new int[N];
    hi = new int[N];
    lo = new int[N];
    poss = new ArrayList<ArrayList<Integer>>();
    intervals = new TreeSet<Interval>();

    for (int i = 0; i < N; i++)
      val[i] = readInt();

    for (int i = 0; i < 2 * N; i++)
      poss.add(new ArrayList<Integer>());

    Stack<State> s = new Stack<State>();

    // processing upper bound (first number less than the current number)
    for (int i = N - 1; i >= 0; i--) {
      while (!s.isEmpty() && val[i] < s.peek().val)
        s.pop();
      if (s.isEmpty())
        hi[val[i]] = N;
      else
        hi[val[i]] = s.peek().index;
      s.push(new State(val[i], i));
    }

    s.clear();
    //processing lower bound (last number greater than the current number)
    for (int i = 0; i < N; i++) {
      while (!s.isEmpty() && val[i] > s.peek().val)
        s.pop();
      if (s.empty())
        lo[val[i]] = -1;
      else
        lo[val[i]] = s.peek().index;
      s.push(new State(val[i], i));
    }

    for (int i = 0; i < N; i++) {
      int diff = val[i] - i + N - 1;
      poss.get(diff).add(i);
    }

    for (int i = 0; i < 2 * N; i++) {
      sweep(poss.get(i));
    }

    out.println(intervals.size());
    for (Interval i : intervals)
      out.printf("%d %d\n", i.l + 1, i.r + 1);

    out.close();
  }

  static void sweep (ArrayList<Integer> indexes) {
    PriorityQueue<Event> pq = new PriorityQueue<Event>();
    for (int i = 0; i < indexes.size(); i++) {
      pq.offer(new Event(lo[val[indexes.get(i)]] + 1, 1, indexes.get(i)));
      pq.offer(new Event(hi[val[indexes.get(i)]], -1, indexes.get(i)));
    }

    TreeSet<Integer> active = new TreeSet<Integer>();
    while (!pq.isEmpty()) {
      Event curr = pq.poll();
      if (curr.type == 1)
        active.add(curr.index);
      else if (curr.type == -1) {
        active.remove(curr.index);
        Integer lower = active.lower(curr.index);
        if (lower != null && lower > lo[val[curr.index]]) {
          Interval add = new Interval(lower, curr.index);
          Interval prev = intervals.floor(add);
          Interval next = intervals.ceiling(add);
          boolean intersectPrev = true;
          boolean intersectNext = true;

          if (prev != null) {
            if (Math.max(add.l, prev.l) < Math.min(add.r, prev.r)) {
              if (add.r - add.l <= prev.r - prev.l) {
                intervals.remove(prev);
                intersectPrev = false;
              }
            } else {
              intersectPrev = false;
            }
          } else {
            intersectPrev = false;
          }

          if (next != null) {
            if (Math.max(add.l, next.l) < Math.min(add.r, next.r)) {
              if (add.r - add.l <= next.r - next.l) {
                intervals.remove(next);
                intersectNext = false;
              }
            } else {
              intersectNext = false;
            }
          } else {
            intersectNext = false;
          }

          if (!intersectPrev && !intersectNext)
            intervals.add(add);
        }
      }
    }
  }

  static class State {
    int val, index;

    State (int val, int index) {
      this.val = val;
      this.index = index;
    }
  }

  static class Event implements Comparable<Event> {
    int x, index, type;

    Event (int x, int type, int index) {
      this.x = x;
      this.type = type;
      this.index = index;
    }

    @Override
    public int compareTo (Event e) {
      if (x == e.x && type == e.type)
        return e.index - index;
      if (x == e.x)
        return type - e.type;
      return x - e.x;
    }
  }

  static class Interval implements Comparable<Interval> {
    int l, r;

    Interval (int l, int r) {
      this.l = l;
      this.r = r;
    }

    @Override
    public int compareTo (Interval i) {
      return l - i.l;
    }
  }

  static String next () throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong () throws IOException {
    return Long.parseLong(next());
  }

  static int readInt () throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble () throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter () throws IOException {
    return next().charAt(0);
  }

  static String readLine () throws IOException {
    return br.readLine().trim();
  }
}
